Sources/Testing/Traits/AttachmentSavingTrait.swift

```diff
@@ -0,0 +1,337 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+/// A type that defines a condition which must be satisfied for the testing
+/// library to save attachments recorded by a test.
+///
+/// To add this trait to a test, use one of the following functions:
+///
+/// - ``Trait/savingAttachments(if:)``
+///
+/// By default, the testing library saves your attachments as soon as you call
+/// ``Attachment/record(_:named:sourceLocation:)``. You can access saved
+/// attachments after your tests finish running:
+///
+/// - When using Xcode, you can access attachments from the test report.
+/// - When using Visual Studio Code, the testing library saves attachments to
+///   `.build/attachments` by default. Visual Studio Code reports the paths to
+///   individual attachments in its Tests Results panel.
+/// - When using Swift Package Manager's `swift test` command, you can pass the
+///   `--attachments-path` option. The testing library saves attachments to the
+///   specified directory.
+///
+/// If you add an instance of this trait type to a test, any attachments that
+/// test records are stored in memory until the test finishes running. The
+/// testing library then evaluates the instance's condition and, if the
+/// condition is met, saves the attachments.
+@_spi(Experimental)
+public struct AttachmentSavingTrait: TestTrait, SuiteTrait {
+  /// A type that describes the conditions under which the testing library
+  /// will save attachments.
+  ///
+  /// You can pass instances of this type to ``Trait/savingAttachments(if:)``.
+  public struct Condition: Sendable {
```

Comment:
My biggest feedback on this PR and associated proposal is that I think there should be a streamlined way to express "save attachments _unconditionally_", whether the test passed or failed. That can be expressed in the current implementation using `.savingAttachments(if: true)` but it feels pretty indirect.

Maybe zero-arg `.savingAttachments()` (or `.savesAttachments()`)?

Reply:
That's the default behaviour, so the streamlined spelling is "".

Reply:
Discussed with Stuart and we'll defer this question to a separate PR.